delta-lake/src/main/java/org/apache/iceberg/delta/BaseMigrateDeltaLakeTableAction.java

```diff
@@ -0,0 +1,356 @@
+    List<Action> actions = versionLog.getActions();
+
+    // We first need to iterate through to see what kind of transaction this was. There are 3
+    // cases:
+    // 1. AppendFile - when there are only AddFile instances (an INSERT on the table)
+    // 2. DeleteFiles - when there are only RemoveFile instances (a DELETE where all the records
+    // of file(s) were removed
+    // 3. OverwriteFiles - when there are a mix of AddFile and RemoveFile (a DELETE/UPDATE)
+
+    // Create a map of Delta Lake Action (AddFile, RemoveFile, etc.) --> List<Action>
+    Map<String, List<Action>> deltaLakeActionMap =
+        actions.stream()
+            .filter(action -> action instanceof AddFile || action instanceof RemoveFile)
+            .collect(Collectors.groupingBy(a -> a.getClass().getSimpleName()));
+
+    List<DataFile> filesToAdd = Lists.newArrayList();
+    List<DataFile> filesToRemove = Lists.newArrayList();
+    for (Action action : Iterables.concat(deltaLakeActionMap.values())) {
+      DataFile dataFile = buildDataFileFromAction(action, transaction.table());
+      if (action instanceof AddFile) {
+        filesToAdd.add(dataFile);
+      } else if (action instanceof RemoveFile) {
+        filesToRemove.add(dataFile);
+      } else {
+        throw new ValidationException(
+            "The action %s's is unsupported", action.getClass().getSimpleName());
+      }
+    }
+
+    if (filesToAdd.size() > 0 && filesToRemove.size() > 0) {
+      // Overwrite_Files case
+      OverwriteFiles overwriteFiles = transaction.newOverwrite();
+      filesToAdd.forEach(overwriteFiles::addFile);
+      filesToRemove.forEach(overwriteFiles::deleteFile);
+      overwriteFiles.commit();
+    } else if (filesToAdd.size() > 0) {
+      // Append_Files case
+      AppendFiles appendFiles = transaction.newAppend();
+      filesToAdd.forEach(appendFiles::appendFile);
+      appendFiles.commit();
+    } else if (filesToRemove.size() > 0) {
+      // Delete_Files case
+      DeleteFiles deleteFiles = transaction.newDelete();
+      filesToRemove.forEach(deleteFiles::deleteFile);
+      deleteFiles.commit();
+    }
+  }
+
+  public DataFile buildDataFileFromAction(Action action, Table table) {
+    PartitionSpec spec = table.spec();
+    String path;
+    Map<String, String> partitionValues;
+
+    if (action instanceof AddFile) {
+      AddFile addFile = (AddFile) action;
+      path = addFile.getPath();
+      partitionValues = addFile.getPartitionValues();
+    } else if (action instanceof RemoveFile) {
+      RemoveFile removeFile = (RemoveFile) action;
+      path = removeFile.getPath();
+      partitionValues = removeFile.getPartitionValues();
+    } else {
+      throw new IllegalStateException(
+          String.format(
+              "Unexpected action type for Delta Lake: %s", action.getClass().getSimpleName()));
+    }
+
+    String fullFilePath = getFullFilePath(path, deltaLog.getPath().toString());
+
+    if (partitionValues == null) {
```

Comment:
sounds like partitionValues is null is always invalid, which means we should throw IllegalArgumentException?